### PR TITLE
Fallback mechanism for resource selection

### PR DESCRIPTION
**Pull request by wpdebruin**
New version of this pull after resourceService rewrite to script

This PR adds a fallback mechanism for selecting resource files. Resources can be in the form
lang_COUNTRYCODE_extrainfo
which means a resource named myResource will only be loaded if myResource_lang_COUNTRYCODE_extrainfo.properties is a valid path.
This PR falls back to less selective property files in this order:

myResource_lang_COUNTRYCODE_extrainfo.properties
myResource_lang_COUNTRYCODE.properties
myResource_lang.properties
myResource.properties
myResource_defaultlocale.properties

This PR is quite useful because it makes localization of the cbvalidation module a lot more reliable and easier.

### CODE REVIEW

File: .project [hunk 1]
```diff
@@ -0,0 +1,11 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (lmajano) [inline .project:1]: Can you remove this file @wpdebruin it is no longer needed

File: models/ResourceService.cfc [hunk 2]
```diff
@@ -228,33 +228,41 @@ component singleton accessors="true" {
...
-			var keys = rb.getKeys();
-
-			// Loop through property keys and store the values into bundle
-			while ( keys.hasMoreElements() ) {
-				thisKEY                   = keys.nextElement();
-				resourceBundle[ thisKEY ] = rb.handleGetObject( thisKEY );
-			}
-		} finally {
-			fis.close();
+		// try to load hierarchical resource rbfile_LANG_COUNTRY_VARIANT, start with default resource, followed by
```
Comment (lmajano) [inline models/ResourceService.cfc:231]: Can you apply the formatting so we can have spacing correctly applied.  Thanks!

File: models/ResourceService.cfc [hunk 3]
```diff
@@ -490,4 +498,37 @@ component singleton accessors="true" {
...
+		try{
+			//get keys
+			var keys = rb.getKeys();
+			//Loop through property keys and store the values into bundle
+			while( keys.hasMoreElements() ){
+				thisKey = keys.nextElement();
+				resourceBundle[ thisKey ] = rb.handleGetObject( thisKey );
+			}
+		}
+		catch( any e ) {
```
Comment (lmajano) [inline models/ResourceService.cfc:525]: I would say remove the catch and add a `finally` block, and this way you can consolidate the `fis.close()` into a single finally block

File: models/ResourceService.cfc [hunk 4]
```diff
@@ -228,33 +228,41 @@ component singleton accessors="true" {
...
+			smartBundleFiles.append("#myRbFile#.properties");	
+		}, "_" );
+		
+		// load all resource files for all default,lang, country and variants 
+		// and overwrite parent keys when present so you you will always have defaults
+		// AND specific resource values for countries and variants without duplicating everything.
+		var resourceBundle = structNew();
+		var IsValidBundleLoaded = false;
+		smartBundleFiles.each(function(resourceFile){
+			var ResourceBundleFullPath = variables.controller.locateFilePath( resourceFile );
```
Comment (lmajano) [inline models/ResourceService.cfc:254]: variables should be lowercased camel case: `resourceBundleFullPath`

File: models/ResourceService.cfc [hunk 5]
```diff
@@ -228,33 +228,41 @@ component singleton accessors="true" {
...
+			if ( ResourceBundleFullPath.len() ) {
+				resourceBundle.append( _loadSubBundle( ResourceBundleFullPath ), true ); //append and overwrite
+				IsValidBundleLoaded = true; // at least one bundle loaded so no errors
```
Comment (lmajano) [inline models/ResourceService.cfc:257]: variables should be lowercased camel case: `isValidBundleLoaded`

File: models/ResourceService.cfc [hunk 6]
```diff
@@ -228,33 +228,41 @@ component singleton accessors="true" {
...
+			};
+		});
+
+		// Validate resource is loaded or error.
+		if( !IsValidBundleLoaded ){
+			var rbFilePath = "#arguments.rbFile#_#arguments.rbLocale#.properties";
+			var rbFullPath = variables.controller.locateFilePath( rbFilePath );
+			throw("The resource bundle file: #rbFilePath# does not exist. Please check your path", "FullPath: #rbFullPath#", "ResourceBundle.InvalidBundlePath");
```
Comment (lmajano) [inline models/ResourceService.cfc:265]: Use named arguments please, better, for visibility: `message, type, details`

Comment (lmajano) [reply models/ResourceService.cfc:265]: Also, the method annotations should have a `@throws InvalidBundlePath`

**Review by lmajano — CHANGES_REQUESTED** [review]

File: models/ResourceService.cfc (hunk 3 above)
Comment (wpdebruin) [reply models/ResourceService.cfc:525]: Ok done. I thought we needed it for a rethrow but you are right. Only a finally block will be sufficient

File: models/ResourceService.cfc (hunk 6 above)
Comment (wpdebruin) [reply models/ResourceService.cfc:265]: Ok fixed named arguments for throw in more places and added @throws in more places (getResourceFileInputStream can throw this)

**Comment by wpdebruin** [comment]
should we move  the config to modulesettings in coldbox config? That would mean more consistent (and easier) code for retrieving settings in the module. And it would also mean more consistency for end users when configuring cbi18n.
Second thought: cbi18n is setting some cookies. Should we replace this by the cbstorages module? Might be more flexible, but adds a new dependency for this module.
Third: should we spent time to created db based storage for resources? I will never use it, because it makes versioning a lot harder in a DTAP flow. But maybe other people like it, so they can change resources on the fly in some CMS.

**Comment by lmajano** [comment]
>should we move the config to modulesettings in coldbox config? That would mean more consistent (and easier) code for retrieving settings in the module. And it would also mean more consistency for end users when configuring cbi18n.

Yes, just note, that this would then trigger a major update as we will break compat! But I am ok with that.

>Second thought: cbi18n is setting some cookies. Should we replace this by the cbstorages module? Might be more flexible, but adds a new dependency for this module.

Hmm, true.I think it would be great to migrate to cbstorages, since now we have a big case statment for storage. This way it can be more flexible. again, this promites a major bump as well. Great idea!

>Third: should we spent time to created db based storage for resources? I will never use it, because it makes versioning a lot harder in a DTAP flow. But maybe other people like it, so they can change resources on the fly in some CMS.

Well, since we made it flexible so you can plug in your own bundle service. I think this is a minor feature we can add if asked enough

**Comment by wpdebruin** [comment]
> > should we move the config to modulesettings in coldbox config? That would mean more consistent (and easier) code for retrieving settings in the module. And it would also mean more consistency for end users when configuring cbi18n.
> 
> Yes, just note, that this would then trigger a major update as we will break compat! But I am ok with that.

Are there many modules depending on cbi18n? I think cbvalidation which also should be updated anyway. But others?
If you are OK with breaking compat I will create this update.
> 
> > Second thought: cbi18n is setting some cookies. Should we replace this by the cbstorages module? Might be more flexible, but adds a new dependency for this module.
> 
> Hmm, true.I think it would be great to migrate to cbstorages, since now we have a big case statment for storage. This way it can be more flexible. again, this promites a major bump as well. Great idea!

OK, will do that as well
> 
> > Third: should we spent time to created db based storage for resources? I will never use it, because it makes versioning a lot harder in a DTAP flow. But maybe other people like it, so they can change resources on the fly in some CMS.
> 
> Well, since we made it flexible so you can plug in your own bundle service. I think this is a minor feature we can add if asked enough

OK, not creating this right now.

**Comment by wpdebruin** [comment]
There are a few minor issues left
dontloadRBFlag in setFwLocale is not used at all, so why is it there?
Shouldn't we make cookiestorage a default? Saves a lot of configuration issues for some people. Or do you want people to be aware where their current locale is stored by explicitly specifying it?

And now I can continue with the JSON language resources.

**Comment by lmajano** [comment]
- `dontloadRBFlag` I don't remember, we would have to verify if it is in use.
- `setFwLocale` is the main function to set the locale, we would not be able to get rid of this function.

>Shouldn't we make cookiestorage a default? Saves a lot of configuration issues for some people. Or do you want people to be aware where their current locale is stored by explicitly specifying it?

Yes, I am ok with this. I actually thought it was like this before.

File: ModuleConfig.cfc [hunk 7]
```diff
@@ -7,200 +7,106 @@
...
 	 * Configure Module
 	 */
-	function configure(){
-		// cb4 style
-		arrayAppend( controller.getSetting( "ApplicationHelper" ), "#moduleMapping#/helpers/Mixins.cfm" );
+	function configure() {
+		settings = {
+			"defaultResourceBundle" : "",
+			"defaultLocale"         : "",
+			"localeStorage"         : "",
```
Comment (lmajano) [inline ModuleConfig.cfc:31]: You can add `cookie` here like you mentioned

File: ModuleConfig.cfc [hunk 8]
```diff
@@ -7,200 +7,106 @@
...
-					localeStorage = "",
-					unknownTranslation = "",
-					logUnknownTranslation = false,
-					resourceBundles = {},
-					customResourceService = ""
+				// set defaults
+				modules[ thisModule ].cbi18n = {
+					defaultResourceBundle : "",
+					defaultLocale         : "",
+					localeStorage         : "",
```
Comment (lmajano) [inline ModuleConfig.cfc:75]: Same here

File: models/i18n.cfc [hunk 9]
```diff
@@ -35,15 +38,34 @@ component singleton accessors="true" {
...
-		variables.defaultLocale         = variables.controller.getSetting( "DefaultLocale" );
-
-		// set locale setup on configuration file
+		variables.localeStorage         = variables.settings.LocaleStorage;
+		variables.defaultResourceBundle = variables.settings.DefaultResourceBundle;
+		variables.defaultLocale         = variables.settings.DefaultLocale;
+
+		// instantiate storage service for locale storage
+		//		try {
+		variables.storageService = wirebox.getInstance(
```
Comment (lmajano) [inline models/i18n.cfc:53]: I don't think it's necessary for you to send in the `initArguments` to get the right storage object.  This should be done automatically by cbStorages.  All we need to do, is request the right storage object according to the `localeStorage` setting.

**Review by lmajano — CHANGES_REQUESTED** [review]

File: ModuleConfig.cfc (hunk 8 above)
Comment (wpdebruin) [reply ModuleConfig.cfc:75]: done

File: models/i18n.cfc (hunk 9 above)
Comment (wpdebruin) [reply models/i18n.cfc]: Yes I see now. A bit hidden in the annotations of the storages init functions :-)

**Pull request merged**